### PR TITLE
ICU-747,ICU-748/refactor internal/oplog and internal/perms to use domain errors

### PR DESCRIPTION
**Pull request by louisruch**
### What does this PR do

Refactors internal/oplog to use the new domain errors
Refactors internal/perms to use the new domain errors

### CODE REVIEW

**Review by talanknight — DISMISSED** [review]
This looks great!
One thing that feels a little strange but I think is ok to move forward with is the "TicketAlreadyRedeemed" and "TicketNotFound" codes seem very specific to a single location where all the other codes are more generally useful.  Again, I don't think this is a problem, just something that stands out a little to me.

File: internal/errors/code.go [hunk 1]
```diff
@@ -57,4 +62,6 @@ const (
 	ColumnNotFound       Code = 1102 // ColumnNotFound represent that a column was not found in the underlying db
 	MaxRetries           Code = 1103 // MaxRetries represent that a db Tx hit max retires allowed
 	Exception            Code = 1104 // Exception represent that an underlying db exception was raised
+
+	FixMe Code = 11111
```
[the remark this answers is not in the document]
Comment (louisruch) [reply internal/errors/code.go:66]: HAHA, thanks this is my solution to avoiding context switching.  I found that when doing the refactor if I come across an error that does not immediately have a code that comes out to me I mark it with this and move on :/

**Comment by louisruch** [comment]
> 
> 
> This looks great!
> One thing that feels a little strange but I think is ok to move forward with is the "TicketAlreadyRedeemed" and "TicketNotFound" codes seem very specific to a single location where all the other codes are more generally useful. Again, I don't think this is a problem, just something that stands out a little to me.

Yeah this is fair I did not want to reuse `NotFound` (db not found) and I felt reusing `KeyNotFound` did not match so I made a new one but yes it is very specific for this use.  With regards to `TicketAlreadyRedeemed`, I did need to match on this in a number of places in the repo, I could go back to having an `ErrTicketAlreadyRedeemed` but I did not have a good code for this error anyway.

**Review by talanknight — DISMISSED** [review]

**Review by jimlambrt — CHANGES_REQUESTED** [review]
some minor bits.  looks great!

File: internal/oplog/queue.go [hunk 2]
```diff
@@ -94,32 +96,36 @@ func (q *Queue) Add(m proto.Message, typeName string, t OpType, opt ...Option) e
...
 	if err != nil {
-		return nil, 0, nil, nil, err // intentionally not wrapping error so client can test for sentinel EOF error
+		return nil, 0, nil, nil, errors.New(errors.Io, op, "binary read error", errors.WithWrap(err))
 	}
 	data := q.Next(int(n))
 	msg := new(AnyOperation)
 	err = proto.Unmarshal(data, msg)
 	if err != nil {
-		return nil, 0, nil, nil, fmt.Errorf("error marshaling the msg for Remove: %w", err)
+		return nil, 0, nil, nil, errors.New(errors.Decode, op, "error unmarshalling message", errors.WithWrap(err))
```
[the remark this answers is not in the document]
Comment (louisruch) [reply internal/oplog/queue.go]: Changed it to one `l` but my non-US education would disagree with you :p 
https://en.wikipedia.org/wiki/Marshalling_(computer_science)

**Pull request merged**